src/Common/tests/System/Collections/IEnumerable.Generic.Tests.cs

```diff
@@ -0,0 +1,742 @@
+                {
+                    while (enumerator.MoveNext()) ;
+                    ModifyEnumerable(enumerable, modificationCode);
+                    Assert.Throws<InvalidOperationException>(() => enumerator.MoveNext());
+                }
+            });
+        }
+
+        [Fact]
+        public void IEnumerable_Generic_Enumerator_MoveNextHitsAllItems()
+        {
+            RepeatTest(
+                (enumerator, items) =>
+                {
+                    var iterations = 0;
+                    while (enumerator.MoveNext())
+                    {
+                        iterations++;
+                    }
+                    Assert.Equal(items.Length, iterations);
+                });
+        }
+
+        [Fact]
+        public void IEnumerable_Generic_Enumerator_MoveNextFalseAfterEndOfCollection()
+        {
+            RepeatTest(
+                (enumerator, items) =>
+                {
+                    while (enumerator.MoveNext())
+                    {
+                    }
+
+                    Assert.False(enumerator.MoveNext());
+                });
+        }
+
+        #endregion
+
+        #region Enumerator.Current
+
+        [Fact]
+        public void IEnumerable_Generic_Enumerator_CurrentThrowsAfterEndOfCollection()
+        {
+            if (Enumerator_Current_UndefinedOperation_Throws)
+            {
+                RepeatTest(
+                    (enumerator, items) =>
+                    {
+                        while (enumerator.MoveNext())
+                        {
+                        }
+                        Assert.Throws<InvalidOperationException>(
+                            () => enumerator.Current);
+                    });
+            }
+        }
+
+        [Fact]
+        public void IEnumerable_Generic_Enumerator_Current()
+        {
+            // Verify that current returns proper result.
+            RepeatTest(
+                (enumerator, items, iteration) =>
+                {
+                    if (iteration == 1)
+                    {
+                        VerifyEnumerator(
+                            enumerator,
+                            items,
+                            0,
+                            items.Length / 2,
+                            true,
+                            false);
+                    }
+                    else
+                    {
+                        VerifyEnumerator(enumerator, items);
+                    }
+                });
+        }
+
+        [Theory]
+        [MemberData("ValidCollectionSizes")]
+        public void IEnumerable_Generic_Enumerator_Current_FromStartToFinish(int count)
+        {
+            using (IEnumerator<T> enumerator = GenericIEnumerableFactory(count).GetEnumerator())
+            {
+                T current;
+                while (enumerator.MoveNext())
+                    current = enumerator.Current;
+            }
```

Comment:
Correctness it tested in  ```IEnumerable_Generic_Enumerator_Current``` test. This test is just to check for exceptions. Not really necessary though since we have the more robust validation already.